### DISCUSSION TITLE
Renderer Backend for Tk

### DISCUSSION THREAD

**Discussion by nico-robert** [ocornut/imgui#5576] [New Users Build/Link/Run issues ONLY!]
Hello, 
I couldn’t find any discussion about that. 
I would like to know if you think it would be possible to integrate `Dear Imgui` in [Tk](https://en.wikipedia.org/wiki/Tk_(software)) window ?
If yes, would this file `imgui_impl_win32.cpp` be a good starting point ?

Regards
Nicolas

**Comment by rokups** [comment]
What you are looking for is how to mix Tk and graphics views (OpenGL/DirectX). There is [this example](https://codeplea.com/opengl-with-c-and-tcl-tk) showing how to display OpenGL scene in a Tk application. You could use `imgui_impl_opengl3.cpp`+`imgui_impl_win32.cpp` to set up OpenGL scene and display it in a Tk window as shown in linked tutorial.

**Reply by nico-robert** [reply]
@rokups I will look at this. Thanks

**Comment by nico-robert** [comment]
I got there...
Now the problem I have, is that the mouse and keyboard are disabled (no problem with `imgui_impl_opengl3.cpp` + `imgui_impl_glfw.cpp`) but the focus works (it’s strange...)
I think _Tk_ window takes control.
Do you have any idea what’s going on ?
![aGpvvpMCun](https://user-images.githubusercontent.com/77193946/186107626-9d0f2aad-0e4a-4b94-9361-0cbceed7a4e5.gif)

**Reply by rokups** [reply]
Look what happens with events, whether they are correctly passed to Dear ImGui. Debug `ImGui_ImplWin32_WndProcHandler` in `imgui_impl_win32.cpp`.

**Comment by nico-robert** [comment]
If I understand , and if I follow this example [main.cpp](https://github.com/ocornut/imgui/blob/master/examples/example_win32_directx9/main.cpp) (from examples/example_win32_directx9)
`ImGui_ImplWin32_WndProcHandler` is a callback ? (I'm using cimgui bindings and noobs in c++ ) but I see that `WndProc` is declared in a `WNDCLASSEX struct` to be included to ` ::CreateWindow`.
In my case my window is not created with Windows API . 
Do you think it can work the same way ?

**Reply by sergeyn** [reply]
You use ImGui_ImplWin32_WndProcHandler as an example which demonstrates what to do with ImGui based on occurrence of a particular win32 UI event

**Reply by ocornut** [reply]
You should be able to obtain the HWND from TK, get the current WndProc handler, hook yours instead and call the original.

**Reply by ocornut** [reply]
Or reimplement its content.

**Reply by edilanejaderlan-cyber** [reply]
Hello I'm i developer i like ImGui bundle and ImGui

**Comment by nico-robert** [comment]
Ok , thanks all for the advice ,here my values :
![eALIf143FV](https://user-images.githubusercontent.com/77193946/186370482-5487abad-28e8-4936-bb2c-498622921912.gif)
What I can see , Once my frame is moved, I have my values that don’t change :
`msg: 0xF (15) wParam: 0 lParam: 0`

**Reply by nico-robert** [reply]
I think I found... I had no callback return with `DefWindowProc`. It seems to work.
Thanks all

**Reply by nico-robert** [reply]
I have another problem, with the keyboard activation, mouse works weel , but the keyboard is not actived , I have read this : [How can I enable keyboard or gamepad controls ? ](https://skia.googlesource.com/external/github.com/ocornut/imgui/+/refs/heads/master/docs/FAQ.md#q-how-can-i-enable-keyboard-or-gamepad-controls)
As always no problem with `imgui_impl_opengl3.cpp` + `imgui_impl_glfw.cpp`  but with my Tk window , yes.
It’s always related to WndProc handler ? In some examples I can have some enumerations used WM_KEYDOWN, WM_KEYUP WM_SYSKEYDOWN... In my case I never see these values.

**Reply by nico-robert** [reply]
[SetFocus](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setfocus) solved my problem
